Sources/StreamChatUI/Utils/AudioPlayer/AudioPlayerObserving.swift

```diff
@@ -0,0 +1,132 @@
+//
+// Copyright © 2023 Stream.io Inc. All rights reserved.
+//
+
+import AVFoundation
+import Foundation
+
+/// Describes an object that can provide AVPlayer related updates.
+public protocol AudioPlayerObserving {
+    /// Registers an observer with the player that will receive updates when the player's
+    /// ``timeControlStatus`` gets updated.
+    /// - Parameters:
+    /// player: The player from which we would like to receive updates
+    /// block: The block to call once a ``timeControlStatus`` update occurs
+    func addTimeControlStatusObserver(
+        _ player: AVPlayer,
+        using block: @escaping (AVPlayer.TimeControlStatus?) -> Void
+    )
+
+    /// Registers an observer that will periodically invoke the given block during playback to
+    /// report changing time.
+    /// - Parameters:
+    /// player: The player from which we would like to receive updates
+    /// interval: The interval at which we would like to receive updates
+    /// queue: The queue on which the call to block will occur every time there is an update
+    /// block: The block to call every time there is an update
+    func addPeriodicTimeObserver(
+        _ player: AVPlayer,
+        forInterval interval: CMTime,
+        queue: DispatchQueue?,
+        using block: @escaping () -> Void
+    )
+
+    /// Registers and observer that will be called once the playback of an item stops
+    /// - Parameters:
+    /// queue: The queue on which the ``stoppedPlaybackHandler`` will be dispatched
+    /// block: The block to call once a player's item has stopped
+    func addStoppedPlaybackObserver(
+        queue: OperationQueue?,
+        using block: @escaping (AVPlayerItem) -> Void
+    )
+}
+
+open class StreamPlayerObserver: AudioPlayerObserving {
+    /// An observer token that is used to observe the ``timeControlStatus`` of the player to receive
+    /// updates when the state changes between playing and paused.
+    private var timeControlStatusObserver: NSKeyValueObservation?
+
+    /// A token referencing the periodicTimer that is registered on the player and is used to provide
+    /// time-related metadata updates.
+    private var periodicTimeObservationToken: Any?
```

Comment:
Curious why are these `Any`?

Reply:
Those are KVO/AVFoundation observation tokens (like subscription receipts on Combine) and the frameworks are actually returning `Any`. As their aim is to provide a strong reference to the observer so it won't be deallocated, we don't need to interact with them, just hold them. Once the token is deallocated, the observer will also be deallocated, invalidating the observation flow.